src/ucp/am/eager_multi.c

```diff
@@ -146,3 +146,144 @@ ucp_proto_t ucp_eager_am_bcopy_multi_proto = {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return ucp_proto_multi_init(&params, params.super.super.priv,
+                                params.super.super.priv_size);
+}
+
+static UCS_F_ALWAYS_INLINE size_t ucp_proto_am_eager_multi_zcopy_add_payload(
+        ucp_request_t *req, const ucp_proto_multi_lane_priv_t *lpriv,
+        ucp_datatype_iter_t *next_iter, uct_iov_t *iov)
+{
+    size_t user_hdr_size = req->send.msg_proto.am.header_length;
+    size_t meta_size, max_payload;
+
+    if (req->send.state.dt_iter.offset == 0) {
+        meta_size = UCP_AM_FIRST_FRAG_META_LEN + user_hdr_size;
+    } else {
+        meta_size = UCP_AM_MID_FRAG_META_LEN;
+    }
+    max_payload = ucp_proto_multi_max_payload(req, lpriv, meta_size);
+
+    return ucp_datatype_iter_next_iov(&req->send.state.dt_iter, max_payload,
+                                      lpriv->super.md_index,
+                                      UCP_DT_MASK_CONTIG_IOV, next_iter, iov,
+                                      lpriv->super.max_iov - 1);
+}
+
+static UCS_F_ALWAYS_INLINE void ucp_proto_am_eager_multi_zcopy_add_footer(
+        ucp_request_t *req, const ucp_proto_multi_lane_priv_t *lpriv,
+        uct_iov_t *iov, size_t *iov_count)
+{
+    size_t user_hdr_size = req->send.msg_proto.am.header_length;
+    size_t footer_size, footer_offset;
+    ucp_am_first_ftr_t *first_ftr;
+
+    if (req->send.state.dt_iter.offset == 0) {
+        footer_size   = sizeof(ucp_am_first_ftr_t) + user_hdr_size;
+        footer_offset = 0;
+        first_ftr     = UCS_PTR_BYTE_OFFSET(req->send.msg_proto.am.reg_desc + 1,
+                                            user_hdr_size);
+        /* The method also fills middle/last fragment footer. The footer can be
+           reused by all fragments because it is immutable. */
+        ucp_am_fill_first_footer(first_ftr, req);
+    } else {
+        footer_size   = sizeof(ucp_am_mid_ftr_t);
+        footer_offset = user_hdr_size;
+    }
+    ucp_proto_eager_am_zcopy_add_footer(req, footer_offset, lpriv->super.lane,
+                                        iov, iov_count, footer_size);
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_am_fill_middle_header(ucp_am_mid_hdr_t *hdr, ucp_request_t *req)
+{
+    hdr->offset = req->send.state.dt_iter.offset;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_proto_am_eager_multi_zcopy_send_func(
+        ucp_request_t *req, const ucp_proto_multi_lane_priv_t *lpriv,
+        ucp_datatype_iter_t *next_iter)
+{
+    union {
+        ucp_am_hdr_t     first;
+        ucp_am_mid_hdr_t middle;
+    } hdr;
+    ucp_am_id_t am_id;
+    size_t hdr_size, iov_count;
+    uct_iov_t iov[UCP_MAX_IOV];
+
+    iov_count = ucp_proto_am_eager_multi_zcopy_add_payload(req, lpriv,
+                                                           next_iter, iov);
+    ucp_proto_am_eager_multi_zcopy_add_footer(req, lpriv, iov, &iov_count);
+
+    if (req->send.state.dt_iter.offset == 0) {
```

Comment:
there are 3 places that check (req->send.state.dt_iter.offset == 0), can we consolidate them?